### ISSUE TITLE
Issue with iOS Service Extension and notifee_options when app in foreground

### ISSUE THREAD

**Issue by ccorneliusHsv** [notifee/react-native-notifee#339]
### Issue: 
Looking to add images to our iOS notifications. I was able to utilize the service extension as documented here -https://notifee.app/react-native/docs/ios/remote-notification-support. Using the FCM Console to send test messages. 
Everything worked fine till I tried modify _bestAttemptContent_  to add the **notifee_options**. It works properly while the app is closed\background. When the app is in the Foreground, the notification is not shown. I did see where "content-available" is required to fire onMessage when app is in foreground, but onMessage did fire with the FCM Message just as it was sent from the testing console. I did test adding (test 2) with adding "content-available" in the same fashion as I added "notifee_options"...but maybe that's not valid 🤷🏼‍♂️  
Details on testing, messages sent, and variations below:

### Setup
   Message sent via Firebase Test Console(derived from self.bestAttemptContent.userInfo in the service extension call: 
```
{
   aps :  {
      alert : {
         body : "Shop Now!",
         title : "Back to School TECH Deals Are Here"
      },
      "mutable-content" : 1
   },
   "fcm_options" :  {
      image : "https://fanart.tv/fanart/movies/15596/movieposter/back-to-school-568913d00c872.jpg"
   }
}
```

###     Test 1 - Notifee Service Extension: 
      
 Service extension code with Notifee changes

```
 - (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];

  [NotifeeExtensionHelper populateNotificationContent:self.bestAttemptContent withContentHandler:contentHandler];
}
```

### Results : 
 
-  Background\Dead: Service Extension is called, Shows notification - no image
-   Foreground: Service Extension is called, messaging().onMessage() is called, displays notification correctly via use of notifee.displayNotification 

### Test 2 - Notifee as written in the guide with modified message from FCM to include notifee_options and content-avaialble: 

Object passed to populateNotificationContent
```
{
   aps :  {
      alert : {
         body : "Shop Now!"
         title : "Back to School TECH Deals Are Here"
      },
      "mutable-content" : 1,
      "content-available" : 1
   },
   "fcm_options" :  {
      image : "https://fanart.tv/fanart/movies/15596/movieposter/back-to-school-568913d00c872.jpg"
   },
   "notifee_options" :  {
      image : "https://fanart.tv/fanart/movies/15596/movieposter/back-to-school-568913d00c872.jpg"
   }
}
```

Code used to modify object: 

```
- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
  NSDictionary *fcmOptions = self.bestAttemptContent.userInfo[@"fcm_options"];
  NSString *imageUrl = fcmOptions[@"image"];
  
  NSMutableDictionary *mutableUserInfo = [self.bestAttemptContent.userInfo mutableCopy];
  
  //Need to add notifee_options: {image: imageUrl} to user info
  NSString *imageKey =@"image";
  NSDictionary * imageDict =@{imageKey:imageUrl};
  
  [mutableUserInfo setValue:imageDict forKey:@"notifee_options"];
  
  //Need to add 'content-available': 1, // Important, to receive `onMessage` event in the foreground when message is incoming
  NSMutableDictionary *mutableAps = [self.bestAttemptContent.userInfo[@"aps"] mutableCopy];
  NSNumber *numWthType = [NSNumber numberWithInt:1];
  [mutableAps setValue:numWthType forKey:@"content-available"];
  [mutableUserInfo setValue:mutableAps forKey:@"aps"];

  self.bestAttemptContent.userInfo = [NSDictionary dictionaryWithDictionary:mutableUserInfo];
  NSLog( @"UserInfo after => '%@'", self.bestAttemptContent.userInfo );

  [NotifeeExtensionHelper populateNotificationContent:self.bestAttemptContent withContentHandler:contentHandler];

```

### Results: 
-  Background\Dead: Service Extension is called, Shows notification with image
-  Foreground: Service Extension is called, Notification is not displayed

  
### Test 3 - Using self.contentHandler(self.bestAttemptContent); instead of notifee handler modified message from FCM to include 

In order to test if my code was an issue, I went back to the following code for the service extension. 

```
- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];

  self.contentHandler(self.bestAttemptContent);
}
```

### Results: 
-  Background\Dead: Service Extension is called, Shows notification - no image
-  Foreground: Service Extension is called, messaging().onMessage() is called, displays notification correctly via use of notifee.displayNotification  

    

Notifee version: 1.9.2

info Fetching system and libraries information...
System:
    OS: macOS 11.4
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 1.70 GB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.3.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.0.14 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.10.0 - /Users/ccornelius/.rvm/gems/ruby-2.7.2/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
    Android SDK: Not Found
  IDEs:
    Android Studio: 4.1 AI-201.8743.12.41.6953283
    Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
  Languages:
    Java: 15.0.1 - /Library/Java/JavaVirtualMachines/jdk-15.0.1.jdk/Contents/Home/bin/javac
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1 
    react-native: 0.63.4 => 0.63.4 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

**Comment by helenaford** [comment]
hey there, thanks for the info - I've had a look through and done some tests. When the app is in the foreground, you have to use `displayNotification` inside `messaging().onMessage`:
```
  await notifee.displayNotification({
    title: 'hello',
    body: 'world',
    ios: {
      attachments: [{ url: 'https://fanart.tv/fanart/movies/15596/movieposter/back-to-school-568913d00c872.jpg'}]
    },
  });
```

**Comment by ccorneliusHsv** [comment]
@helenaford I agree with the first part of your response and have it implemented just like you described. 

But, while I expected the behavior to match your statement "when the app is in the foreground, the notification service extension is not called"..that is not the behavior I am seeing. I am using the Console app to look at the logs and when the app is in the foreground and receives the push, the service extension is going through the same steps as when the app is closed or in the background.  

Also - just for clarity `messaging().onMessage`  is called when the app is in the foreground, after the service extension is called, if I don't have the `notifee_options` set.  

Let me know if there is any additional testing I can do or information I can provide.

**Comment by helenaford** [comment]
@ccorneliusHsv can you summarize in some bullet points the issue you're having and steps on how to recreate it? Just so I'm clear what the bug is

**Comment by ccorneliusHsv** [comment]
- Issue: `messaging().onMessage` is not being called when app is in Foreground
- I am modifying the incoming message inside didReceiveNotificationRequest  to include `notifee_options` (that code is in the original post directly under Code used to modify object)
- If I don't modify the incoming message to include `notifee_options`, then `messaging().onMessage` is called.

**Comment by helenaford** [comment]
Ah ok, got it! thanks. I can 100% say `onMessage` should be called but the fact you say it breaks when you're modifying the incoming message in the service extension is interesting. Also, I will check again but I'm still not sure why the extension will be getting called in the foreground 🤔 In test 1 you say it works, can you not modify it on the client? Is there a reason you need to pass fcm_options too?

**Comment by ccorneliusHsv** [comment]
Yes,  `fcm_options` are required..at least for the message sent to the device. Believe it or not our marketing team uses the FCM console to send an occasional message..and that is why that's there. I will try removing the `fcm_options` in the service extension when I add the `notifee_options`.  

We also send other notifications and I'm currently working on modifying the format to get the Service Extension to handle it.  When I get that working, I might have some more information for you.

**Comment by ccorneliusHsv** [comment]
Did some more testing..I was able to send a message via the Firebase API that did not include `fcm_options`..observed the same behavior as with the other messages.
